inmoose/diffexp/meta.py

```diff
@@ -0,0 +1,93 @@
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+import numpy as np
+import pandas as pd
+from scipy.stats import combine_pvalues, false_discovery_control
+from statsmodels.stats.meta_analysis import combine_effects
+
+from .DEResults import DEResults
+
+
+def meta_de(de_results, alpha=0.05):
+    """
+    Combine logFC and *p*-values of differential expression analyses
+
+    log-fold-changes are combined using a random-effect model, where the random
+    effects variance is iteratively estimated with the Paule-Mandel method.
+    Confidence intervals for the combined log-fold-change values are computed
+    assuming a normal distribution and without scaling.
+
+    *p*-values are combined using Fisher's combined probability test, then
+    adjusted for multiple testing with Benjamini-Hochberg procedure.
+
+    Arguments
+    ---------
+    de_results : list of DEResults
+        the list of differential expression results to combine. Depending on
+        the use-case, it can be results obtained with different tools on the
+        same dataset, results obtained with the same tool on different
+        datasets, or any combination thereof
+    alpha : float between 0 and 1
+        significance level for the confidence intervals. Defaults to 0.05.
+
+    Returns
+    -------
+    pd.DataFrame
+        a dataframe indexed by genes with the following columns:
+
+        - :code:`"combined logFC"`: the combined log-fold-change
+        - :code:`"combined logFC (CI_L)"`: the lower bound of the confidence
+          interval for the combined log-fold-change
+        - :code:`"combined logFC (CI_R)"`: the lower bound of the confidence
+          interval for the combined log-fold-change
+        - :code:`"adjusted combined pval"`: the combined *p*-value, adjusted
+          for multiple testing
+    """
+    if len(de_results) < 2:
+        raise ValueError("metaanalysis requires at least two analyses to aggregate")
+    if not all(isinstance(d, DEResults) for d in de_results):
+        raise ValueError(
+            "diffexp results fed to metanalysis must inherit from DEResults"
+        )
+    idx = de_results[0].index
+    if not all(np.array_equal(d.index, idx) for d in de_results):
+        raise ValueError(
+            "diffexp results fed to metaanalysis must have the same set of genes"
+        )
```

Comment:
Couldn't we provide an option to do the gene intersection and apply meta-analysis? We can have a minimum intersection length maybe.